src/client/lazy-app/Modal/index.tsx

```diff
@@ -0,0 +1,120 @@
+import { h, Component, VNode, Fragment } from 'preact';
+import * as style from './style.css';
+import 'add-css:./style.css';
+import { linkRef } from 'shared/prerendered-app/util';
+import { cleanSet } from '../util/clean-modify';
+import { animateFrom, animateTo } from '../util';
+import { RoundedCrossIcon } from '../icons';
+
+interface Props {
+  icon: VNode;
+  title: string;
+  content: VNode;
+}
+
+interface State {}
+
+export default class Modal extends Component<Props, State> {
+  private dialogElement?: HTMLDialogElement;
+
+  componentDidMount() {
+    if (!this.dialogElement) throw new Error('Modal missing');
+  }
+
+  private _closeOnTransitionEnd = () => {
+    // If modal does not exist
+    if (!this.dialogElement) return;
+
+    this.dialogElement.close();
+  };
+
+  showModal() {
+    if (!this.dialogElement) throw Error('Modal missing');
+    if (this.dialogElement.open) return;
+
+    this.dialogElement.showModal();
+    // animate modal opening
+    animateTo(
+      this.dialogElement,
+      [
+        { opacity: 0, transform: 'translateY(50px)' },
+        { opacity: 1, transform: 'translateY(0px)' },
+      ],
+      { duration: 250, easing: 'ease' },
+    );
+    // animate modal::backdrop
+    // some browsers don't support ::backdrop, catch those errors
+    try {
+      animateFrom(
+        this.dialogElement,
+        { opacity: 0 },
+        {
+          duration: 250,
+          easing: 'ease',
+          pseudoElement: '::backdrop',
+        },
+      );
+    } catch (e) {}
+  }
+
+  private _hideModal() {
+    if (!this.dialogElement) throw Error('Modal missing / hidden');
+    if (!this.dialogElement.open) return;
+
+    // animate modal closing
+    const anim = animateTo(
+      this.dialogElement,
+      { opacity: 0, transform: 'translateY(50px)' },
+      { duration: 250, easing: 'ease' },
+    );
+    // animate modal::backdrop
+    // some browsers don't support ::backdrop, catch those errors
+    try {
+      animateTo(this.dialogElement, [{ opacity: 0 }], {
+        duration: 250,
+        easing: 'ease',
+        pseudoElement: '::backdrop',
+      });
+    } catch (e) {}
+    anim.onfinish = this._closeOnTransitionEnd;
```

Comment:
```suggestion
    anim.finished.then(() => this._closeOnTransitionEnd());
```

Using the promise makes it clearer that we only expect this to happen once. Also, it avoids passing an event into a function that doesn't expect it. https://jakearchibald.com/2021/function-callback-risks/